serializer, README: Fix a minor issue and add usage examples

romainletendart opened the pull request.

Commit:
README.rst: Add (de)serialization examples

Commit:
structure: Let OCPPMessage subclasses define their messageTypeId

lburg commented on README.rst:
```diff
@@ -1,8 +1,64 @@
 *ocpp-codec* provides dataclasses definitions of OCPP messages and types, in both version 1.6 and 2.0 of the protocol.
 
-It also provides a JSON encoder to exchange messages.
+It also provides a JSON pre-serializer and post-deserializer to exchange messages (see below).
```
Should be `pre-encoder` and `post-decoder` here to match the following sections

lburg commented on ocpp_codec/structure.py:
```diff
@@ -45,6 +45,9 @@ class Call(OCPPMessage):
     payload: dict
 
+    def __post_init__(self):
+        self.messageTypeId = MessageTypeEnum.CALL
```
Since the field was defined as `init=False`, you can simply assign it  at the class level:

```python
class Call:
    messageTypeId = MessageTypeEnum.CALL
```

lburg commented on ocpp_codec/structure.py:
```diff
@@ -32,7 +32,7 @@ class ErrorCode(types.SimpleType):
 class OCPPMessage:
     """Base class every OCPP message should inherit from."""
-    messageTypeId: MessageType
+    messageTypeId: MessageType = field(init=False)  # Let subclasses define that field
```
I wonder if we shouldn't simply remove it altogether from the dataclass fields... Like not defining it in `OCPPMessage` at all, and only define it class-wide on `Call`, `CallResult` and `CallError`, without type annotation.

This way, in the `parse_structure` method, we can simply skip it from `raw_data` (`raw_data = raw_data[1:]`) after having used it to get the dataclass we need through `_MSGTYPEID_TO_DATACLASS`.

What do you think? I don't really like the `__post_init__` methods and using `del`.

rsebille replied:
I think we should keep it here, we shouldn't split the structure between the dataclasses and the logical code.
`messageTypeId` will always be overridden but at least when someone jump on board it will be clear that it's one of the two fields required in every OCPP message.

And like you said, `__post_init__` shouldn't be needed if `messageTypeId` is made a class member.

romainletendart replied:
I chose the `__post_init__` way because of a [warning from python's documentation](https://docs.python.org/3.7/library/dataclasses.html?highlight=__post_init#dataclasses.replace) about `init=False` fields:

> Be forewarned about how init=False fields work during a call to replace(). They are not copied from the source object, but rather are initialized in __post_init__(), if they’re initialized at all. It is expected that init=False fields will be rarely and judiciously used. If they are used, it might be wise to have alternate class constructors, or perhaps a custom replace() (or similarly named) method which handles instance copying.

With a second thought though, one should not alter an instance's `messageTypeId` and calling `replace()` on such altered message will simply use the default `messageTypeId` for the new instance which is what we want anyway.

lburg commented on ocpp_codec/serializer.py:
```diff
@@ -400,7 +400,10 @@ def serialize_fields(message):
+            if issubclass(field.type, types.ComplexType):
+                serialize_func = serialize_fields
+            else:
+                serialize_func = functools.partial(serialize_field, field)
```
I don't really understand the problem this is solving. You seem to add a restriction on `optionalComplexListValue` number of elements, but then in your test you test only the validation of the elements it contains?

romainletendart replied:
I've split the commit in two and reworded both commit messages.
Could you tell me if it's clearer now? :)

lburg reviewed (commented):
I think commit messages could provide more information about the problems they solve!

rsebille commented on tests/test_serializer.py:
```diff
@@ -471,6 +479,13 @@ def test_serialize_fields():
+        # List of complex types holding unvalidated data
+        serializer.serialize_fields(messages.ComplexAction.conf(
+            optionalComplexListValue=[
+                types.ComplexType(enumValue=types.FooBarEnum.Foo, validatedValue=100 * 'a'),
```
```suggestion
                types.ComplexType(enumValue=types.FooBarEnum.Foo, validatedValue='a' * 100),
```
;)

Commit:
ocpp_codec: Fix serialization of List[non-ComplexType] fields

`serialize_field` takes 2 arguments but `serialize_func` only provides
one in case of List[non-ComplexType].

Commit:
tests: Test validation on serialization of List[ComplexTypes] fields

lburg commented on tests/messages.py:
```diff
@@ -33,6 +33,9 @@ class req:
     class conf:
         optionalListValue: typing.List[str] = None
+        optionalComplexListValue: typing.List[types.ComplexType] = field(
+            default=None, metadata={'validator': validators.max_length_4},
```
A test is missing to validate a list max length, but it's a good idea!

lburg commented:
:clap:
